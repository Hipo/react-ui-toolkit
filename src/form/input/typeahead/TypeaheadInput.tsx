import React, {useEffect} from "react";
import classNames from "classnames";

import Input from "../Input";
import useDebounce from "../../../core/utils/hooks/debounce";

export interface TypeaheadInputProps {
  testid?: string;
  customClassName?: string;
  id?: string;
  name: string;
  isDisabled?: boolean;
  value?: string;
  placeholder: string;
  queryChangeDebounceTimeout?: number;
  onQueryChange: (value: string) => void;
  onFocus?: React.ReactEventHandler<HTMLInputElement>;
  onBlur?: React.ReactEventHandler<HTMLInputElement>;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  role?: string;
  children?: React.ReactNode;
  inputContainerRef?: React.RefObject<HTMLDivElement>;
  shouldResetValue?: boolean;
}

const DEFAULT_DEBOUNCE_TIMEOUT = 250;

function TypeaheadInput(props: TypeaheadInputProps) {
  const {
    testid,
    placeholder,
    name,
    customClassName,
    onFocus,
    onBlur,
    id,
    role,
    onKeyDown,
    onQueryChange,
    isDisabled = false,
    leftIcon,
    rightIcon,
    value = "",
    queryChangeDebounceTimeout = DEFAULT_DEBOUNCE_TIMEOUT,
    inputContainerRef,
    shouldResetValue
  } = props;

  const [inputValue, setInputValue] = useDebounce(
    onQueryChange,
    value,
    queryChangeDebounceTimeout
  );

  useEffect(() => {
    if (shouldResetValue) {
      setInputValue("");
    }
  }, [shouldResetValue, setInputValue]);

  return (
    <Input
      inputContainerRef={inputContainerRef}
      customClassName={classNames("typeahead-input", customClassName)}
      id={id}
      testid={testid}
      name={name}
      placeholder={placeholder}
      onChange={handleInputChange}
      onFocus={onFocus}
      onBlur={onBlur}
      onKeyDown={onKeyDown}
      value={inputValue}
      isDisabled={isDisabled}
      leftIcon={leftIcon}
      rightIcon={rightIcon}
      role={role}
    />
  );

  function handleInputChange(event: React.SyntheticEvent<HTMLInputElement>) {
    setInputValue(event.currentTarget.value);
  }
}

export default TypeaheadInput;
