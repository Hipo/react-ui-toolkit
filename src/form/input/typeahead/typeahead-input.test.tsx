import React from "react";
import {render} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import {testA11y} from "../../../core/utils/test/testUtils";
import TypeaheadInput, {TypeaheadInputProps} from "./TypeaheadInput";

describe("<TypeaheadInput />", () => {
  const defaultTypeaheadInputProps: TypeaheadInputProps = {
    testid: "typeahead-input",
    name: "typeahead-input",
    onQueryChange: jest.fn(),
    placeholder: "typeahead input"
  };

  it("should render correctly", () => {
    render(<TypeaheadInput {...defaultTypeaheadInputProps} />);
  });

  it("should match snapshot", () => {
    const tree = create(<TypeaheadInput {...defaultTypeaheadInputProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<TypeaheadInput {...defaultTypeaheadInputProps} />);

    await testA11y(container);
  });
});
