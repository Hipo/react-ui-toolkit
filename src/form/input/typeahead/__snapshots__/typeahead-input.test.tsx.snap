// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TypeaheadInput /> should match snapshot 1`] = `
<div
  className="input-container typeahead-input"
  data-testid="typeahead-input"
>
  <input
    autoComplete="off"
    autoCorrect="off"
    className="input"
    disabled={false}
    name="typeahead-input"
    onChange={[Function]}
    placeholder="typeahead input"
    type="text"
    value=""
  />
</div>
`;
